interactions/utils/abc/base_iterators.py

```diff
@@ -0,0 +1,97 @@
+from abc import ABCMeta, abstractmethod
+from math import inf
+from typing import TYPE_CHECKING, Callable, List, Optional, TypeVar, Union
+
+from ..missing import MISSING
+
+_T = TypeVar("_T")
+_O = TypeVar("_O")
+
+__all__ = ("BaseAsyncIterator", "BaseIterator")
+
+if TYPE_CHECKING:
+    from ...api.http.client import HTTPClient
+    from ...api.models.misc import Snowflake
+
+
+class BaseAsyncIterator(metaclass=ABCMeta):
+    """A base class for async iterators."""
+
+    # I don't want to make it subclass the BaseIterator since it forces implementation of __next__ and __iter__
+
+    @abstractmethod
+    def __init__(
+        self,
+        obj: Union[int, str, "Snowflake", _T],
+        _client: Optional["HTTPClient"] = None,
```

Comment:
Can you make it so that obj, client, etc. aren't implemented in the base async iterator? There are use cases where none of the parameters will be needed

Reply:
I can make them none but I don't want to remove them since the objects are intended to be used internally hence the design, it just saves like 20-30 lines of duplicate code

Reply:
It also prevents them from being used internally for purposes other than pagination from discord

Reply:
How so if they're optional?